[addons] fixes: disable pinning for system addons / gui usability improvements

howie-f opened the pull request:
## Description
### Commit 1:
system addons, for instance `service.xbmc.versioncheck`, `metadata.themoviedb.org` should not be auto pinned on an update

### Commit 2:
corrects the number of available addon-updates displayed on the home-screen by introducing infolabel `AddonUpdateCount`
(thx @ronie)

### Commit 3:
`DialogAddonInfo`: improve usability. the `Versions` button moves to the right and turns into an `Update` button, if an update is available `AND` the local outdated addon has `Auto-updates` enabled aka is not pinned.
The available Update-Version is also displayed in the dialog now.

![screenshot00001](https://user-images.githubusercontent.com/58829855/95728023-47337f00-0c7b-11eb-8eed-826ec5fda7ff.png)

![screenshot00003](https://user-images.githubusercontent.com/58829855/95728041-4e5a8d00-0c7b-11eb-83d1-aae37dbc88a5.png)

## How Has This Been Tested?
debian stretch local

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [X] **Improvement** (non-breaking change which improves existing functionality)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document

howie-f committed:
[addons] no auto version pinning for ORIGIN_SYSTEM addons

howie-f committed:
[addons] introduce combined update/versions-button for AddonInfoDialog

AlwinEsch commented on addons/skin.estuary/xml/DialogAddonInfo.xml:
```diff
@@ -245,6 +245,7 @@
 						<param name="id" value="12" />
 						<param name="icon" value="icons/infodialogs/launch.png" />
 						<param name="label" value="$LOCALIZE[518]" />
+						<param name="visible" value="!ListItem.Property(Addon.IsBinary)" />
```
For now OK, but have in mind that also it can come that binary addon becomes runway to have independent. Maybe also some other addon types not need it.

There should be the most correct to have e.g. Addon.IsRunnable, to have not fixed to a language and where checks his type.

AlwinEsch replied:
Also not sure if them hidden that GUI his flow then match the others. Maybe only button disabled and not hidden 🤔 .

@Razzeee what you thing?

Razzeee replied:
Disabling would be better, but I might be missing some context here. I checked Leia and it seemed like we already did disable this button? Was that removed in between?

![image](https://user-images.githubusercontent.com/5943908/95675919-f78d7e80-0bba-11eb-93fc-a6fb91064eac.png)

howie-f replied:
> Disabling would be better, but I might be missing some context here. Was that removed in between?

agreed and i would disable it in core, not the skin. think there hasn‘t been removed something in between but i also cannot „run“ `pvr.hts` for instance in leia

Commit:
[addons] correct displayed number of available updates in gui

garbear commented:
But binary add-ons can be run? The Game API has a `supports_standalone` flag to allow for standalone "running": https://github.com/kodi-game/game.libretro.mrboom/blob/master/game.libretro.mrboom/addon.xml.in#L15

howie-f commented:
@garbear disabling the run-button intended to remove its availability after installation of `NextPVR` under windows, but i think it's better to drop this commit for now until we have a properly working `Addon.IsRunnable`

garbear reviewed (approved):
Looks good. Notes:

* There's some whitespace irregularity in the addons/skin.estuary/xml/Variables.xml addition
* Are `Addon.xxx` listitem properties documented anywhere? I remember making a forum thread for all game-related listitem properties but I don't think we centralize documentation for anything else

howie-f commented:
fixed the whitespace and also changed `!String.IsEmpty(ListItem.Property(Addon.UpdateTargetVersion))` to `ListItem.Property(Addon.HasUpdate)`.. just simpler. I'm not aware of any `Addon.Property` documentation, but sure would catch up if someone points me to it.

ronie commented:
these are documented in GUIInfoManager.cpp, starting here:
https://github.com/xbmc/xbmc/blob/b0dfc080b153a30f9af7e2d45abf230c30ead201/xbmc/GUIInfoManager.cpp#L5846

ronie commented on addons/skin.estuary/xml/DialogAddonInfo.xml:
```diff
@@ -217,7 +217,7 @@
 								<item>
 									<label>$LOCALIZE[126]:</label>
-									<label2>$INFO[ListItem.Property(Addon.Status)]</label2>
+									<label2>$VAR[AddonStatus]</label2>
```
you can accomplish this without a `$VAR[]` if you use the `$INFO[infolabel,prefix,postfix]` format.
this format will hide the prefix & postfix if the infolabel returns empty.

```
$INFO[ListItem.Property(Addon.Status)]$INFO[ListItem.Property(Addon.UpdateTargetVersion),[CR]($LOCALIZE[19114]: ,)]
```

howie-f replied:
done + the requested doc

ronie commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -4488,6 +4497,12 @@ const infomap container_str[]  = {{ "property",         CONTAINER_PROPERTY },
+///   \table_row3{   <b>`ListItem.Property(Addon.ValidUpdateVersion)`</b>,
+///                  \anchor ListItem_Property_ValidUpdateVersion
+///                  _string_,
+///     @return The version string of a valid update for the addon. Empty string if there is no valid update available.
```
please insert the 'new skinning label' bit:
```
///     <p><hr>
///     @skinning_v19 **[New Infolabel]** \link ListItem_Property_ValidUpdateVersion `ListItem.Property(Addon.ValidUpdateVersion)`\endlink
```

garbear commented on xbmc/filesystem/AddonsDirectory.cpp:
```diff
@@ -826,12 +826,20 @@ void CAddonsDirectory::GenerateAddonListing(const CURL& path,
+    std::string validUpdateVersion;
+    if (hasUpdate)
+    {
+      const auto& mapEntry = addonsWithUpdate.find(addon->ID());
```
std::map returns a lightweight iterator, instead of a reference to the underlying object. It doesn't impact performance to drop the `&`. This might also be incorrect, as the iterator might be destructed after this line, and the reference is unsafely dangling at reclaimable memory.

Also, the constness applies to the underlying object, not the iterator, so `const` can be dropped too. The iterator propagates constantness itself without needing to be const.

howie-f replied:
makes sense to do this, i'm wondering if the same goes here
https://github.com/xbmc/xbmc/blob/a816c1759a38410528c53acd208c163fb84c1475/xbmc/filesystem/AddonsDirectory.cpp#L817
and it should better be `std::shared_ptr<IAddon> checkedObject =`

phunkyfish replied:
Passing by reference is ok with `shared_ptr`.

garbear replied:
Agreed. The problem is when the object can be dereferenced elsewhere. I actually fixed a crash where a shared_ptr was passed by reference, and the reference count dropped to zero elsewhere. Fixed by passing by value adding another refcount, `std::shared_ptr<IAddon> checkedObject =`. But here reference is fine because it's so short lived, not worth the atomic ref count increment

howie-f replied:
yeah, that sounds good. thanks for the explanation :)

howie-f commented:
@ronie are we good with the actual state?

ronie reviewed (approved):
looks good to me.
